polynote-frontend/polynote/ui.js

```diff
@@ -918,32 +927,45 @@ export class NotebookUI {
             }
         });
 
-        socket.addMessageListener(messages.UpdateCell, (path, globalVersion, localVersion, id, edits) => {
-            if (globalVersion >= this.globalVersion) {
-                this.globalVersion = globalVersion;
-                let rebasedEdits = edits;
+        socket.addMessageListener(messages.NotebookUpdate, update => {
+            if (update.path === this.path) {
+                if (update.globalVersion >= this.globalVersion) {
+                    this.globalVersion = update.globalVersion;
 
-                if (localVersion < this.localVersion) {
-                    const editsBefore = this.editBuffer.range(localVersion, this.localVersion);
-                    rebasedEdits = messages.ContentEdit.rebaseEdits(edits, editsBefore);
-                } else if (localVersion > this.localVersion) {
-                    console.error("Local version from server is after client's. Something's wrong.")
-                }
+                    if (update.localVersion < this.localVersion) {
+                        const prevUpdates = this.editBuffer.range(update.localVersion, this.localVersion);
+                        update = messages.NotebookUpdate.rebase(update, prevUpdates);
+                    }
 
-                this.localVersion++;
 
-                const cell = this.cellUI.getCell(id);
+                    this.localVersion++;
+
+                    match(update)
+                        .when(messages.UpdateCell, (p, g, l, id, edits) => {
+                            const cell = this.cellUI.getCell(id);
+                            if (cell) {
+                                cell.applyEdits(edits);
+                                this.editBuffer.push(this.localVersion, messages);
+                            }
+                        })
+                        .when(messages.InsertCell, (p, g, l, cell, after) => {
```

Comment:
Another one for the TODO list: I think it's inevitable that inserting a cell on the client will have to block until the server acknowledges, so that the server can assign the ID. Otherwise if you and I both insert a cell at the same time, they'll have conflicting IDs and catastrophe of some kind must ensue.

Reply:
That's one way - perhaps an alternative would be that once the server gets both insertions it could linearize and rebase the cell IDs in the same way it handles edits to cells? Whichever client wins gets the ID it assigns, and the other client assigns a new one and tries again (or maybe at this point the server can give it a new one... depends on how many clients might be adding cells at a given time :-\ )

I wonder, though, whether it might be useful to have a concept of passive and active clients with only one active at a time. 

That way:
- I can look at your notebook without worrying about messing things up.
- We can simplify coordination by coordinating a single lock. 
- Passive clients only need to accept updates which might make reconciliation easier since we don't have to worry about blowing away local state.
- Passive clients can take the lock at any time and become active with a simple consensus of clients: passive clients agree to any leadership change, the current active client either agrees or times out. (once we have user auth then we can allow only authorized users to activate the notebook). 

This might not be worthwhile - depends on how close we are to fully coordinated concurrent editing (even in that case, passive mode might still be useful but we wouldn't need any locks). But if we start running into coordination problems I wonder whether something like this could help, especially as I'm not sure how much need there is for true concurrent editing as a feature really (rather than a way to prevent things from breaking I mean).